docs/guides/guide-for-writing-plugins.md

```diff
@@ -0,0 +1,81 @@
+
+
+You should check out this [link](https://github.com/astrale-sharp/wasm-minimal-protocol). The repo contains:
+- A list of examples of plugin implementations.
+- A test runner for these examples.
+- Wrappers to help you write your plugin in Rust. (Zig wrappers in development)
+- A stubber for wasi [here](https://github.com/astrale-sharp/wasm-minimal-protocol#you-need-to-stub-a-webassembly-plugin)
+
+
+# Protocol
+
+This section describes the protocol Typst expects plugins to implement. This protocol sends and receive byte slices with Typst as the host.
+
+Types and functions are described using WAT syntax.
+
+## Compilation
+
+This protocol is only meant to be used by plugins compiled to 32-bits WebAssembly.
+
+A plugin should compile to a shared WebAssembly library.
+
+## Imports
+
+Valid plugins need to import two functions (that will be provided by the runtime):
+
+- `(import "typst_env" "wasm_minimal_protocol_write_args_to_buffer" (func (param i32)))`
+
+  The argument is a pointer to a buffer (`ptr`).
+
+  Write the arguments for the current function into the buffer pointed at by `ptr`.
+
+  Each function for the protocol receives lengths as its arguments (see [User-defined functions](#user-defined-functions)). The capacity of the buffer pointed at by `ptr` should be at least the sum of all those lengths.
+
+- `(import "typst_env" "wasm_minimal_protocol_send_result_to_host" (func (param i32 i32)))`
+
+  The first parameter is a pointer to a buffer (`ptr`), the second is the length of the buffer (`len`).
+
+  Send `len` and `ptr` to host memory. The buffer must not be freed by the end of the function: it will be freed by the runtime by calling [`wasm_minimal_protocol_send_result_to_host`](#exports).
+
+  If the message should be interpreted as an error message (see [User-defined functions](#user-defined-functions)), it should be encoded as UTF-8.
+
+### Note
+
+If [`wasm_minimal_protocol_send_result_to_host`](#exports) calls `free` (or a similar routine), be careful that the buffer does not point to static memory.
+
+## Exports
+
+Valid plugins need to export a function named `wasm_minimal_protocol_send_result_to_host`, that has signature `func (param i32 i32)`.
+
+This function will be used by the runtime to free the block of memory returned by a [user-defined](#user-defined-functions) function.
+
+## User-defined functions
+
+To conform to the protocol, an exported function should:
+
+- Take `n` arguments `a₁`, `a₂`, ..., `aₙ` of type `u32` (interpreted as lengths, so `usize/size_t` may be preferable), and return one `i32`. We will call the return `return_code`.
+- The function should first allocate a buffer `buf` of length `a₁ + a₂ + ⋯ + aₙ`, and call `wasm_minimal_protocol_write_args_to_buffer(buf.ptr)`.
+- The `a₁` first bytes of the buffer constitute the first argument, the `a₂` next bytes the second argument, and so on.
```

Comment:
imo some kind of serialization is needed here. current code is very unlikely to work on machines of different endianness. maybe we should use some proper ser/de techniques rather than this "home-made" serilization.

Reply:
From what I can gather, `wasmi` is designed to handle both little and big endian execution environment, so this should not be an issue. I'll ask to get a confirmation.

Reply:
i mean, a 4 byte memory chunk can have different meaning in big/little endian. If just write the in-memory representation of a i32 to somewhere, it can be parsed into a different i32 if the endianess doesn't match. For example if the host use big endian and the wasm plugin use little endian, args will be mis-interpreted. 

A better approach is to reach a consensus in the protocol. For example the port number in a tcp packet is always in big endian and this is called the network endianness.

Reply:
I suggest use proper ser/de techniques because serialization frameworks handle all these problems like struct layout and endianess in the background.

Reply:
True, but we never directly write the representation of those `u32`, we use the abstraction provided by `wasmi`. And it converts everything correctly I believe : that is, it converts arguments to little-endian on the `wasm` side, and convert returns back to native endianness for the host. And `wasm` is standardized to always be little endian.

Or do you mean: what if the plugin itself write some data into the buffer, that may be endian-dependant ? I think this is why we restricted to using bytes: if someone want to pass structured data (say, a `i32`), they **have** to choose an endianness.

Reply:
> we never directly write the representation of those u32, we use the abstraction provided by wasmi. And it converts everything correctly I believe

cool! that makes a lot of sense to me. So seems that it's not a problem now at least for this case.

> I think this is why we restricted to using bytes: if someone want to pass structured data (say, a i32), they have to choose an endianness.

I didn't get why restricting the returning value to bytes can solve this problem. The host always need to parse the bytes in someway to make use of it. So there should be some kind of consensus here to parse the data correctly. Or do we just left that to the writer of the plugin? So the plugin writer should decide how to parse the result.

Reply:
It is like "offload" the ser/de efforts to the plugin author. This is kind of strange to me but it seems to be a good start for now.👍

Reply:
seems resolved then!

Reply:
For the record, [I have confirmation](https://github.com/paritytech/wasmi/issues/749) that endian conversions are automatically done for arguments and results!